[쇼핑 주문 1, 2 단계] 비비 미션 제출합니다.

rosemin928 opened the pull request:
안녕하세요 크롱! 안드로이드 7기 비비입니다 😊
우테코 첫 미션이었던 레이싱카 이후로 우테코 마지막 미션까지 크롱과 함께 하게 되어 정말 영광입니다!!ㅎㅎ
이번에 페어와 저 모두 처음 배우는 것들이 많아서 구현이 많이 미흡합니다...제대로 돌아가지 않는 기능도 많아 추후 수정하도록 하겠습니다!
구현에 신경쓰느라 테스트를 신경쓰지 못해 이 부분도 수정하며 작성하도록 하겠습니다...!!

[이번 미션에 해당하는 코드 커밋입니다!](https://github.com/woowacourse/android-shopping-order/pull/117/files/ff6675cda02665ae4db524b96837f4ffb31f0a8c..66d48d912d01b6f454cc89dcd432923b99502c14)

# 셀프 체크리스트
- [x] 프로그램이 정상적으로 작동하는가?
- [ ] 모든 테스트가 통과하는가?
- [ ] 이전에 받은 피드백을 모두 반영하였는가?
- [x] 코딩 스타일 가이드를 준수하였는가?
    - [x] IDE 코드 자동 정렬을 적용하였는가?
    - [x] 린트 검사를 통과하였는가?
- [x] 프로그래밍 요구 사항을 준수하였는가?
- [x] `README.md`에 기능 목록을 정리하고 명확히 기술하였는가?

# 어떤 부분에 집중하여 리뷰해야 할까요?
- 서버 통신을 위한 직렬화 라이브러리로 `kotlinx serialization`을 사용했습니다! `kotlinx serialization`은 kotlin에 최적화되어 있고, 컴파일 타임에 코드를 생성해 런타임 성능이 뛰어나다는 장점을 가지고 있다고 알고 있어 사용했습니다!
- 사용자 인증 정보를 저장하기 위해 local.properties에
```
//local.properties
USER_ID = rosemin928
USER_PASSWORD = password
BASE_URL = http://techcourse-lv2-alb-974870821.ap-northeast-2.elb.amazonaws.com
```
다음과 같이 적어두었는데, BuildConfig import가 안 되어서 불러올 때는 하드코딩된 코드가 그대로 남아있습니다... 이유를 찾지 못해 그대로 두었는데 이 부분도 추후 수정하겠습니다!!

## 스크린샷
https://github.com/user-attachments/assets/fe69077f-37f0-4e6b-815a-6be25dabf188

적고 나서보니 추후 수정하겠다는 내용이 참 많네요... 리뷰 주시면 열심히 반영하겠습니다!!
크롱의 소중한 시간 내어 읽어주셔서 감사합니다! 🙇‍♀️

rosemin928 committed:
feat: 쇼핑 카트 미션 스쿼시 병합

rosemin928 committed:
test: JSON 학습 테스트 작성

rosemin928 committed:
docs: 1단계 기능명세서 작성

rosemin928 committed:
build: retrofit 의존성 추가

rosemin928 committed:
feat: 상품 목록 불러오는 서비스 인터페이스 작성

rosemin928 committed:
refactor: 패키지 이동

devfeijoa committed:
feat: Retrofit 인스턴스 생성

rosemin928 committed:
feat: 상품 목록 API 응답으로 화면 구현

rosemin928 committed:
feat: 상품 목록 API 응답으로 화면 구현

rosemin928 committed:
Merge remote-tracking branch 'origin/step1' into step1

# Conflicts:
#	app/src/main/java/woowacourse/shopping/feature/goods/GoodsViewModel.kt

rosemin928 committed:
refactor: GoodsViewModel 상품 목록 가져오는 로직 수정

rosemin928 committed:
refactor: 최근 본 상품 불러오는 로직 수정

devfeijoa committed:
refactor: Product에서 Cart로 변경

rosemin928 committed:
feat: 장바구니 불러오기 및 담기 기능 구현

rosemin928 committed:
refactor: 상품 응답 데이터 구조 변경 및 매핑 로직 추가

rosemin928 committed:
refactor: 서버에서 장바구니 목록 불러오도록 수정

rosemin928 committed:
refactor: 서버에서 상품 목록 불러오는 로직 수정

rosemin928 committed:
refactor: 더 이상 불러올 상품이 없을 때 더보기 버튼 안 보이도록 수정

devfeijoa committed:
feat: 각각의 상품 수량 불러오는 기능 구현

devfeijoa committed:
feat: 장바구니 상품 수량 수정 기능 구현

rosemin928 committed:
feat: 장바구니에서 상품 제거하는 기능 구현

devfeijoa committed:
feat: 장바구니에 담긴 상품 개수 조회 기능 구현

rosemin928 committed:
feat: 장바구니에서 상품 완전히 제거하는 기능 구현

rosemin928 committed:
fix: 상품 최초로 담았을 때 더 담기/빼기 안 되는 오류 수정

rosemin928 committed:
refactor: 장바구니에서 상품 담고 빼는 로직 수정

rosemin928 committed:
feat: 장바구니 페이지 기능 구현

devfeijoa committed:
refactor: 상품 상세 조회 기능 구현

rosemin928 committed:
feat: 장바구니에 담긴 수량 보여주는 기능 구현

rosemin928 committed:
docs: 2단계 기능명세서 작성 및 구현된 기능 체크

devfeijoa committed:
refactor: 상품 상세 정보 조회 시 장바구니 정보도 함께 조회하도록 수정

rosemin928 committed:
refactor: 장바구니 상품 목록 fragment로 수정

devfeijoa committed:
refactor: 장바구니 상품 변경사항 서버에 반영

devfeijoa committed:
refactor: 사용하지 않는 함수 제거

rosemin928 committed:
refactor: 상품 목록 스켈레톤 UI 보이도록 수정

rosemin928 committed:
refactor: 상품 상세 페이지에서 장바구니에 담는 로직 수정

rosemin928 committed:
feat: 장바구니 스켈레톤 UI 구현

rosemin928 committed:
fix: 장바구니 페이지 넘길 때 로딩 전에 이전 페이지가 겹쳐보이는 오류 수정

rosemin928 committed:
feat: 장바구니에 담긴 상품 주문 시 추천 상품 화면으로 전환하는 기능 구현

rosemin928 committed:
feat: 장바구니에 최근 본 상품과 동일 카테고리 상품 추천 기능 구현

rosemin928 committed:
fix: 추천 상품 목록에 이미 담긴 상품 제외

rosemin928 committed:
docs: README 기능 요구사항 및 구현 기능 체크박스 업데이트

rosemin928 committed:
feat: BuildConfig에 API 관련 정보 추가 및 Gradle 설정 수정

krrong commented on app/src/main/res/layout/activity_cart.xml:
```diff
@@ -0,0 +1,48 @@
+            android:layout_height="wrap_content"
+            app:layout_constraintTop_toTopOf="parent"
+            app:shimmer_auto_start="true"
+            app:visible="@{viewModel.isLoading}">
```
Shimmer의 가시성을 뷰모델의 값을 통해 제어하셨네요 👍

krrong commented on app/src/main/res/layout/item_goods_shimmer.xml:
```diff
@@ -0,0 +1,42 @@
+        app:layout_constraintStart_toStartOf="parent"
+        app:layout_constraintTop_toTopOf="parent" />
+
+    <TextView
```
Shimmer인데 ImageView, TextView임을 알아야 할 필요가 있을까요?
그냥 View여도 되지 않을까해서요!

rosemin928 replied:
Shimmer의 디자인을 아이템의 디자인과 최대한 같게 하려고 복사해서 사용하는 과정에서 ImageView, TextView가 그대로 들어갔네요 🥲
그냥 View인게 더 깔끔한 것 같아 수정했습니다!

[feat: shimmer UI 개선 및 색상 추가](https://github.com/woowacourse/android-shopping-order/pull/117/commits/23b53c127cd9eb1233a725c92cb7fc0e509055fb)

krrong commented on app/src/main/java/woowacourse/shopping/feature/goods/adapter/GoodsAdapter.kt:
```diff
@@ -0,0 +1,112 @@
+
+    fun setItems(newItems: List<Any>) {
+        val oldHistory = items.firstOrNull()?.takeIf { it is List<*> && it.all { h -> h is Cart } } as? List<*>
+        val newHistory = newItems.firstOrNull()?.takeIf { it is List<*> && it.all { h -> h is Cart } } as? List<*>
```
`GoodsAdapter`는 어디선가 재활용되기는 쉽지 않아보이는데 들어오는 아이템을 Any타입으로 만들어둔 이유가 무엇일까요?

rosemin928 replied:
`GoodsActivity`에서는 어댑터를 하나만 사용하기 때문에, `carts`와 `histories`를 함께 다루기 위해 `items`를 List<Any>로 관리하도록 했습니다. 리뷰를 반영하며 `GoodsItem`이라는 sealed class를 정의하고, items를 List<GoodsItem>로 변경해서 각 타입을 명확하게 처리할 수 있게 수정했습니다!

```kotlin
sealed class GoodsItem {
    data class Recent(
        val histories: List<History>,
    ) : GoodsItem()

    data class Product(
        val cart: Cart,
    ) : GoodsItem()

    object LoadMore : GoodsItem()
}
```

items에 최근 본 상품(Recent), 상품(Product), 더보기 버튼(LoadMore)이 들어가도록 했습니다!

[feat: 상품 화면의 아이템 목록 sealed class로 변경](https://github.com/woowacourse/android-shopping-order/pull/117/commits/f57b6ae7f7323a64c055a6b27b71159810adac7b)
[refactor: 상품 목록 RecyclerView Adapter에 ListAdapter 및 DiffUtil 적용](https://github.com/woowacourse/android-shopping-order/pull/117/commits/7357564555a3a18fc1f2216b7b936a0131926eeb)

krrong replied:
sealed class를 통해 원하는 아이템들만 관리할 수 있도록 변경되었네요 👍

krrong commented on app/src/main/java/woowacourse/shopping/feature/goods/adapter/GoodsAdapter.kt:
```diff
@@ -0,0 +1,112 @@
+                GoodsViewHolder(binding, goodsClickListener)
+            }
+            ItemViewType.LOAD_MORE -> {
+                val binding = ItemLoadMoreBinding.inflate(inflater, parent, false)
```
Adapter와 ViewHolder를 조금 더 객체지향적인 눈으로 바라본다고 하면 ViewHolder의 생성을 Adapter에서 하는 것이 아니라 직접 할 수 있도록 변경할 수 있을 것 같아요!
그러면 어떤 이점이 생길 수 있을까요?

rosemin928 replied:
우선 **SRP**를 지킬 수 있다는 이점이 있을 것 같습니다! Adapter는 어떤 ViewHolder를 만들지만 신경쓰고, 어떻게 만드는지는 ViewHolder가 관리하도록 할 수 있을 것 같아요. 또, 만약 여러 Adapter에서 같은 ViewHolder를 쓴다면 중복 코드를 없앨 수 있다는 이점도 있을 것 같습니다!

[refactor: ViewHolder 생성 로직 companion object로 이동](https://github.com/woowacourse/android-shopping-order/pull/117/commits/b0c3d7653e90c3b3df32dea960fefb0c2ac1cf65)

krrong commented on app/src/main/java/woowacourse/shopping/domain/model/Cart.kt:
```diff
@@ -0,0 +1,9 @@
+package woowacourse.shopping.domain.model
+
+data class Cart(
```
`Cart`라는 네이밍만 가지고서는 아이템인지 알기 어렵겠어요 🤔

rosemin928 replied:
`CartProduct`라는 이름으로 수정해봤습니다!

[refactor: Cart 도메인 모델명을 CartProduct로 변경](https://github.com/woowacourse/android-shopping-order/pull/117/commits/890ee1e1266c19748a81eb18b5786b59d0d7079b)

krrong commented on app/src/main/java/woowacourse/shopping/feature/goods/GoodsActivity.kt:
상품 상세 정보로 들어가서 아이템을 추가하고 다시 돌아온 뒤 같은 아이템을 해당 페이지에서 추가하려고 하면 안되는 것 같아요 😭

https://github.com/user-attachments/assets/5d59ee93-7fb0-45a5-b31c-b7183b1fb928

rosemin928 replied:
상품 목록에서 해당 상품의 `cartId`를 받지 못하고 있어서 생겼던 문제네요...!
`activityResultLauncher`를 통해 `cartId`를 받아 업데이트하도록 해서 수정했습니다!

[fix: 상품 상세페이지에서 수량 변경 시 상품 목록에서 추가/제거가 안 되는 오류 수정](https://github.com/woowacourse/android-shopping-order/pull/117/commits/db2732d5a7e0950515b690bb33f3115d467caf3f)

krrong commented on app/src/main/java/woowacourse/shopping/feature/cart/CartActivity.kt:
장바구니 화면에서 상품을 삭제하는 기능을 개선해보면 좋겠어요!

+ 전체 선택 부분도 확인해보면 좋겠네요

https://github.com/user-attachments/assets/059040b5-8582-4541-be9b-bc35ef7d3d67

rosemin928 replied:
장바구니에서 삭제하면 장바구니 리스트에서도 사라져 보이도록 수정했습니다!
또, 전체 선택 부분도 구현했습니다!

[fix: 장바구니에서 상품 삭제 시 수량 감소 대신 목록에서 제거하도록 수정](https://github.com/woowacourse/android-shopping-order/pull/117/commits/757923439fcd1f20e50127db30a6941697a6d331)
[feat: 장바구니 상품 개별 및 전체 선택/해제 기능 추가](https://github.com/woowacourse/android-shopping-order/pull/117/commits/68ad5f5c63b9a6c56fb4c96c21fc4997ea85b58f)

krrong commented on app/src/main/java/woowacourse/shopping/feature/goods/GoodsViewModel.kt:
```diff
@@ -0,0 +1,251 @@
+    val isFail: SingleLiveData<Unit> get() = _isFail
+    private val _isLoading = MutableLiveData<Boolean>(true)
+    val isLoading: LiveData<Boolean> get() = _isLoading
+    private var page: Int = INITIAL_PAGE
```
개행을 해보는건 어떨까요 😅

rosemin928 replied:
빠르게 구현하려 하다보니 전체적으로 가독성이 많이 떨어지네요 🥲
앞으로 신경 써서 작성하겠습니다!

krrong commented on app/src/main/java/woowacourse/shopping/feature/goods/GoodsViewModel.kt:
nit: 마이너한 내용이지만, 함수의 호출 순서대로 함수를 정리하는 것이 가독성에 많은 도움이 되기도 합니다!

krrong commented on app/src/main/java/woowacourse/shopping/feature/goods/GoodsViewModel.kt:
```diff
@@ -0,0 +1,251 @@
+                            _isLoading.postValue(false)
+                        },
+                        onError = {
+                            Log.e("loadProductsInRange", "상품 API 실패", it)
```
개발이 끝난 후에는 이러한 로그들을 지워주는 것 또한 중요합니다.

krrong commented on app/src/main/java/woowacourse/shopping/feature/goods/GoodsViewModel.kt:
```diff
@@ -0,0 +1,251 @@
+                    listOf(histories) + cartsOnly
+                } else {
+                    listOf(histories) + currentItems
+                }
```
이건 무슨 작업인가요??

rosemin928 replied:
```
- 이미 histories가 있는 경우에는 현재의 histories를 없애고 새로운 histories를 붙인다.
- histories가 없는 경우에는 바로 histories를 붙인다.
```

라는 내용의 코드였는데 sealed class를 이용해 리팩토링하며 해당 로직도 다음과 같이 수정되었습니다!

```kotlin
fun refreshHistoryOnly() {
        historyRepository.getAll { historiesList ->
            val currentItems = _items.value.orEmpty().toMutableList()
            val cartsOnly = currentItems.filterIsInstance<GoodsItem.Product>()
            val updatedItems = mutableListOf<GoodsItem>()
            if (historiesList.isNotEmpty()) {
                updatedItems.add(GoodsItem.Recent(historiesList))
            }
            updatedItems.addAll(cartsOnly)
            _items.postValue(updatedItems)
        }
    }
```

krrong commented on app/src/main/java/woowacourse/shopping/feature/goods/GoodsViewModel.kt:
```diff
@@ -0,0 +1,251 @@
+    private val _isFail = MutableSingleLiveData<Unit>()
+    val isFail: SingleLiveData<Unit> get() = _isFail
+    private val _isLoading = MutableLiveData<Boolean>(true)
+    val isLoading: LiveData<Boolean> get() = _isLoading
```
로딩 상태를 잘 관리해주고 계시네요 👍
각각의 상태를 하나의 라이브데이터로 관리하기보다 sealed class의 형태로 관리한다면 하나의 라이브데이터로 활용할 수 있지 않을까요?

rosemin928 replied:
`isSuccess`와 `isFail`을 하나의 State로 묶어 관리하도록 변경했습니다!
처음에는 `loading`까지 포함해서 sealed class로 묶어보려고 했는데, loading은 화면 전체의 로딩 상태이고, success와 fail은 아이템 삽입이라는 특정 이벤트의 결과 상태라고 판단해 분리를 하게 되었습니다.
그래서 현재는 State에 `Success`와 `Fail`만 포함하고 있고, `loading`은 기존처럼 별도로 관리하고 있는데 이에 대해 크롱은 어떻게 생각하시는지 궁금합니다!

[refactor: 상품 추가 결과 상태 State 클래스로 관리](https://github.com/woowacourse/android-shopping-order/pull/117/commits/92afcea25e24a90ccba69e3bb8461c7b56518c8f)

krrong replied:
장바구니에 담긴 상태와 로딩의 상태는 비비가 말한대로 다른 상태이고, 잘 분리해주신 것 같아요!
이전 리뷰에서 이 라이브데이터들을 보고 전체 화면 데이터 로딩의 성공, 실패, 로딩중 의 상태를 관리하는 것으로 이해했었네요 😛

krrong commented on app/src/main/java/woowacourse/shopping/feature/goods/GoodsViewModel.kt:
```diff
@@ -0,0 +1,251 @@
+    private val _items = MutableLiveData<List<Any>>()
+    val items: LiveData<List<Any>> get() = _items
+    private val products = MutableLiveData<List<Cart>>()
+    private val histories = MutableLiveData<List<Cart>>()
```
외부에서 구독하지 않고 있는 데이터를 라이브 데이터로 만들어주는 이유는 무엇일까용?.?

rosemin928 replied:
굳이 그럴 필요 없을 것 같네요...! 습관적으로 전부 LiveData로 작성한 것 같아요 🥲
수정했습니다!!

[refactor: products와 histories 필드 타입을 MutableList로 변경](https://github.com/woowacourse/android-shopping-order/pull/117/commits/f0742caa5ab7d0cebeeec05d65b87d66c3fc593b)

krrong commented on app/src/main/java/woowacourse/shopping/feature/goods/GoodsViewModel.kt:
```diff
@@ -0,0 +1,251 @@
+                    .onSuccess { response ->
+                        val locationHeader = response.headers()["Location"]
+                        val newId = locationHeader?.substringAfterLast("/")?.toLongOrNull() ?: 0
+                        val updatedCart = cart.copy(id = newId, quantity = newQuantity)
```
이러한 작업들을 뷰모델에서 하는게 적절할지 고민해보면 좋겠어요!

rosemin928 replied:
Repository에서 API 응답 전체(response)를 넘겨주기보다, 필요한 값인 cartId만 넘겨주는 것이 더 명확하고 간결하겠네요!
수정했습니다 😁

```kotlin
fun addToCart(
        cartRequest: CartRequest,
        onResult: (Result<Long>) -> Unit,
) {
        ...
if (response.isSuccessful) {
        val newCartId = response.headers()["Location"]?.substringAfterLast("/")?.toLongOrNull() ?: 0
         onResult(Result.success(newCartId))
        }
        ...
}
```

[refactor: 장바구니 상품 추가 시 반환값을 Response에서 Cart Id(Long)로 변경](https://github.com/woowacourse/android-shopping-order/pull/117/commits/164efe5596d5553c8d4c5958cac1466b9a97cf6c)

krrong commented on app/src/main/java/woowacourse/shopping/feature/goodsdetails/GoodsDetailsActivity.kt:
```diff
@@ -0,0 +1,125 @@
+        setContentView(binding.root)
+
+        id = intent.getLongExtra(GOODS_KEY, 0)
+        viewModel.setInitialCart(id)
```
뷰모델의 생성자로 넘길수도 있지 않을까요?

rosemin928 replied:
뷰모델의 생성자로 id를 넘기는 것이 

```kotlin
class GoodsDetailsViewModel(
    private val id: Long,
    private val cartRepository: CartRepository,
    private val historyRepository: HistoryRepository,
    private val productRepository: ProductRepository,
)
```

이런 식으로 수정해보자는 말씀이 맞나요...??
이렇게 했을 경우에는 GoodsDetailsActivity를 위한 ViewModelFactory를 따로 구현하고, Application에서가 아닌 Activity에서 Factory를 생성해서 조금 더 복잡해질 것 같다는 생각이 드는데, 제가 생각하고 있는 방향이 맞는지 궁금합니다!
만약 아니라면 말씀주시면 다음 리뷰에 다시 반영해보도록 하겠습니다! 🙇‍♀️

krrong replied:
넵 맞습니다!
상세 정보 화면에 들어오고 나서는 뷰모델의 `setInitialCart`함수가 필수적으로 호출이 되어야 하는데요.
이를 잊어버리게 되는 경우 상세 정보에서는 아무런 정보도 보여줄 수 없습니다.

오늘 작성한 코드는 내일의 레거시 대상이고, 쉽게 잊어버리기 마련이에요.
과연 나중에 코드를 리팩터링하는 과정에서도 해당 함수가 필수적으로 불려야하는가를 기억할 수 있을지 고민해봐도 좋을거에요 🤔

ViewModelFactory를 생성하는 방법은 [공식 문서](https://developer.android.com/topic/libraries/architecture/viewmodel/viewmodel-factories?hl=ko)를 참고해봐도 좋을 것 같아요!

krrong commented on app/src/main/java/woowacourse/shopping/feature/goodsdetails/GoodsDetailsActivity.kt:
```diff
@@ -0,0 +1,125 @@
+                putExtra("HISTORY_ID", id)
+            },
+        )
+    }
```
모두 observe하는 일을 하는데 파편화되어 있다는 느낌을 받기도 해요.
observe라는 함수에 함께 위치시키고 만약 해당 함수가 너무 커진다면 observeXxx 처럼 나누어보는건 어떨까요?

rosemin928 replied:
`observeCartInsertResult()`, `observeNavigateToLastViewedCart()`로 나누어주고, onCreate에는 

```kotlin
private fun setupObservers() {
        observeCartInsertResult()
        observeNavigateToLastViewedCart()
    }
```

다음과 같은 함수를 만들어 호출해주도록 수정했습니다!

또, 
```kotlin
setResult(
            ResultCode.GOODS_DETAIL_INSERT.code,
            Intent().apply {
                putExtra("HISTORY_ID", id)
            },
        )
```

이 부분은 observe를 하는 것이 아닌 상품 상세 페이지에 들어오자마자 historyId를 넘겨주는 부분이라 생각해 따로 함수를 만들어 분리해주었습니다!

[refactor: GoodsDetailsActivity 로직 정리](https://github.com/woowacourse/android-shopping-order/pull/117/commits/b1cee11ef943e163bf12522d30a389962b00feaf)

krrong commented on app/src/main/java/woowacourse/shopping/feature/goodsdetails/GoodsDetailsActivity.kt:
```diff
@@ -0,0 +1,125 @@
+    private fun navigateToLastViewed() {
+        viewModel.navigateToLastViewedCart.observe(this) { lastViewedCart ->
+            lastViewedCart.let {
+                val intent = newIntent(this@GoodsDetailsActivity, it.product.id.toLong())
```
Int를 Long으로 강제로 변환시키면 어떤 문제가 발생할 수 있을까요?

rosemin928 replied:
```
- Int -> Long 변환: 제대로 변환해주지 않았을 때 컴파일 오류의 가능성이 있음
- Long -> Int 변환: 데이터 손실이 일어날 가능성이 있음, 마찬가지로 컴파일 오류의 가능성이 있음
```

위와 같은 문제들이 발생할 수 있다고 하네요!
API에서 모든 id 값을 Long으로 사용하고 있어서 동일하게 모두 Long으로 수정해주었습니다 😊

[refactor: Product 및 HistoryEntity id 타입 Long으로 변경](https://github.com/woowacourse/android-shopping-order/pull/117/commits/f55beff33e852c3b734794e89ff458b0fb679175)

krrong commented on app/src/main/java/woowacourse/shopping/feature/goodsdetails/GoodsDetailsActivity.kt:
```diff
@@ -0,0 +1,125 @@
+                ResultCode.GOODS_DETAIL_INSERT.code,
+                Intent().apply {
+                    putExtra("GOODS_ID", id)
+                    putExtra("GOODS_QUANTITY", viewModel.cart.value?.quantity)
```
곳곳에 상수화가 필요한 부분이 보이네요 🤠

rosemin928 replied:
급하게 구현하느라 신경 쓰지 못했네요 😢
전부 상수화해주었습니다!

[refactor: Intent extra 키 값 상수화](https://github.com/woowacourse/android-shopping-order/pull/117/commits/1b6825a4bda39208f1eefa4a6da1859f1286d3df)

krrong commented on app/src/main/java/woowacourse/shopping/feature/goodsdetails/GoodsDetailsViewModel.kt:
```diff
@@ -0,0 +1,194 @@
+    fun loadProductDetails(productId: Long) {
+        cartRepository.fetchAllCart(
+            onSuccess = { response ->
+                productRepository.requestProductDetails(
```
레포지토리에서 상품의 데이터를 가져오고 내부에서 상품의 정보를 조회하도록 하여 반환한다면 인터페이스가 `Product`를 반환하도록 만들수도 있을거에요.

일단 `response`의 형태가 아닌 값을 반환하도록 리팩터링 해본 뒤 해당 내용도 가능하면 적용해보셔도 좋겠어요.

rosemin928 replied:
레포지토리에 `findCartByProductId()` 함수를 추가하여 만약 카트에 해당 product가 있다면 해당 카트의 상세 정보만 가지고 올 수 있도록 했습니다. 이를 사용해 ViewModel에서는 response가 아닌 content만 받아오도록 수정해보았습니다!

[refactor: 상품 상세 조회 시 장바구니에 해당 상품이 있는지 확인하는 로직 추가](https://github.com/woowacourse/android-shopping-order/pull/117/commits/92b952599574fb00e04727912c275a3c4d5f31e7)

krrong commented on app/src/main/java/woowacourse/shopping/feature/goodsdetails/GoodsDetailsViewModel.kt:
```diff
@@ -0,0 +1,194 @@
+        }
+    }
+
+    fun commitCart2() {
```
👀

rosemin928 replied:
앗,, 급하게 여러 함수를 작성해서 테스트해보다 지우지 못하고 남긴 함수가 있네요...ㅎㅎ
해당 함수는 삭제했습니다 🥲

krrong commented on app/src/main/java/woowacourse/shopping/feature/goodsdetails/GoodsDetailsViewModel.kt:
```diff
@@ -0,0 +1,194 @@
+        if (cart.value != null) {
+            val newQuantity = cart.value?.quantity ?: 0
+
+            if (cart.value?.quantity == 1) {
```
앞에서 null 체크를 했지만 여전히 ? 가 필요합니다. 이유가 무엇일까요?

rosemin928 replied:
`cart`가 단순 필드가 아닌 LiveData, 즉 getter를 호출해서 가져오는 값이기 때문에 모든 시점에 null이 아닐거라는 보장이 되지 않아서 `?`이 필요한 것 같습니다!

따라서 현재 시점의 `cart`인 `currentCart`를 만들어 저장하고, 만약 값이 null이면 기본으로 생성한 빈 Cart 값을 사용하도록 수정했습니다!

[refactor: CartProduct.EMPTY_CART_PRODUCT 상수 추가](https://github.com/woowacourse/android-shopping-order/pull/117/commits/7f19977c85b7d906871e402b640a5ab3629a93cb)

krrong commented on app/src/main/java/woowacourse/shopping/feature/goodsdetails/GoodsDetailsViewModel.kt:
```diff
@@ -0,0 +1,194 @@
+                    "",
+                ),
+                cart.quantity,
+            ),
```
필요하다면 적당한 Mapper를 만들어볼 수도 있겠네요

rosemin928 replied:
급하게 구현하면서 insertToHistory()에 Cart를 그대로 넘기고 있었는데, 이제는 History 객체를 생성해서 넘기도록 수정했습니다!

```kotlin
private fun insertToHistory(cart: Cart) {
        historyRepository.insert(
            History(id = cart.product.id, name = cart.product.name, thumbnailUrl = cart.product.imageUrl),
        )
    }
```

바뀐 코드는 Mapper를 만들 만큼 복잡한 변환이 아니라고 판단했고, 대신 id, name, thumbnailUrl을 명시적으로 작성함으로써 들어가는 값의 의도를 코드상에서 바로 파악할 수 있도록 해보았는데 크롱은 어떻게 생각하시나요??

[efactor: History 도메인 모델 생성 및 관련 로직 수정](https://github.com/woowacourse/android-shopping-order/pull/117/commits/5277ec5f215550b477c79f2137b78cbc6356f332)
[refactor: History 객체 생성 시 명명된 인자 사용](https://github.com/woowacourse/android-shopping-order/pull/117/commits/cfa159ab851a122f194992e58e131879040c4146)

krrong replied:
좋은 근거를 가지고 잘 구현해주신 것 같아요 😆

krrong commented on app/src/main/java/woowacourse/shopping/feature/goodsdetails/GoodsDetailsViewModel.kt:
```diff
@@ -0,0 +1,194 @@
+            _lastViewed.postValue(lastViewed)
+            updateLastViewedVisibility()
+        }
+    }
```
최근 본 상품이 넘어왔다는 것으로는 뷰의 가시성을 결정할 수 없는 것일까요?

rosemin928 replied:
최근 본 상품이 넘어왔다는 것으로 뷰의 가시성을 결정한다는 것이

```kotlin
private fun updateLastViewedVisibility() {
        val history = lastViewed.value
        val currentCart = cartProduct.value
        _shouldShowLastViewed.value = history != null && history.name != currentCart?.product?.name
    }
```

 다음과 같이 lastViewed가 null이 아닐 때 보이도록 하라는 말씀이 맞을까요...?
위 코드처럼 lastViewed가 null이 아니고, 현재 상세 페이지의 상품과 최근 본 상품이 같지 않은 경우 보이도록 수정해보았습니다!

```kotlin
private val _shouldShowLastViewed = MediatorLiveData<Boolean>()
val shouldShowLastViewed: LiveData<Boolean> get() = _shouldShowLastViewed

    init {
        _shouldShowLastViewed.addSource(cartProduct) { updateLastViewedVisibility() }
    }
```

또, 액티비티에서 observe 하지 않고, ViewModel 내에서 cartProduct의 값이 바뀔 때 가시성 여부를 판단하도록 수정했습니다!
이렇게 생각하는 방향이 맞을까요...??

[efactor: 최근 본 상품 표시 여부 로직 수정](https://github.com/woowacourse/android-shopping-order/pull/117/commits/3d0e2ac25c11f11aa384c8d0dc386ec9a48ea8ae)

krrong commented on app/src/main/java/woowacourse/shopping/feature/goods/adapter/history/HistoryAdapter.kt:
```diff
@@ -0,0 +1,38 @@
+    fun setItems(newItems: List<Cart>) {
+        items.clear()
+        items.addAll(newItems)
+        notifyDataSetChanged()
```
`notifyDataSetChanged()`함수를 사용하면 IDE가 경고를 주는데요. 이 경고를 주는 이유는 무엇일까요?
`notifyDataSetChanged()`를 사용하고 있는 어댑터를 개선해볼 수 있을까요?

rosemin928 replied:
notifyDataSetChanged()는 RecyclerView 어댑터에서 모든 데이터가 변경되었음을 알리는 메서드입니다. 하지만 이 함수는 데이터의 어떤 항목이 바뀌었는지 구체적으로 알려주지 않기 때문에 전체 리스트를 무조건 다시 그리게 되고, 이런 비효율성 때문에 IDE가 경고를 주는 것으로 알고 있습니다!

이 문제를 해결하기 위해 **ListAdapter**를 사용해 어댑터를 개선했습니다. ListAdapter는 내부적으로 DiffUtil을 사용해 이전 리스트와 새 리스트를 비교하고, 변경된 항목만 효율적으로 업데이트해주기 때문에 사용했습니다!

[refactor: ListAdapter와 DiffUtil을 사용하여 RecyclerView 어댑터 개선](https://github.com/woowacourse/android-shopping-order/pull/117/commits/96a78e1d54fac36dbfab8a67fb074279709f0156)

krrong replied:
ListAdapter로 변경을 해주셨군요!
적지 않은 작업이었을텐데 잘 해주셨네요. 리팩터링을 하시면서 ListAdapter와 RecyclerView Adapter의 차이를 경험해보셨기를 바라요 😊

krrong commented on app/src/main/java/woowacourse/shopping/feature/goods/adapter/history/HistoryViewHolder.kt:
```diff
@@ -0,0 +1,16 @@
+) : RecyclerView.ViewHolder(binding.root) {
+    fun bind(cart: Cart) {
+        binding.cart = cart
+        binding.goodsClickListener = goodsClickListener
```
`bind` 함수가 호출될 때마다 리스너를 바인딩해줘야 할까요?

rosemin928 replied:
`ViewHolder`는 재사용되기 때문에, 클릭 리스너를 init 블록에서 한 번만 설정해두면 매 바인딩마다 다시 설정하지 않아도 되어 재활용 측면에서 더 효율적이라 판단해 수정했습니다!

[refactor: 뷰홀더에서 클릭 리스너를 init 블록에서 한 번만 설정하도록 변경](https://github.com/woowacourse/android-shopping-order/pull/117/commits/d4f8eb956496a9e1a4cb59301af5603db1a973a6)

krrong commented on app/src/main/java/woowacourse/shopping/feature/cart/CartRecommendFragment.kt:
```diff
@@ -0,0 +1,39 @@
+
+class CartRecommendFragment : Fragment() {
+    private lateinit var binding: FragmentCartRecommendBinding
+    private val viewModel: CartViewModel by activityViewModels<CartViewModel>()
```
오 너무 잘 활용해주셨네요 😆

krrong commented on app/src/main/java/woowacourse/shopping/feature/cart/CartProductFragment.kt:
```diff
@@ -0,0 +1,91 @@
+        container: ViewGroup?,
+        savedInstanceState: Bundle?,
+    ): View {
+        binding = DataBindingUtil.inflate(inflater, R.layout.fragment_cart_product, container, false)
```
`DataBindingUtil`을 사용하신 이유가 있을까요?

rosemin928 replied:
```kotlin
// 1번
// ViewBinding
binding = FragmentCartRecommendBinding.inflate(inflater, container, false)

// 2번
// DataBinding
binding = DataBindingUtil.inflate(inflater, R.layout.fragment_cart_recommend, container, false)
```

이라고 생각을 했어서 DataBindingUtil을 사용했는데 다른 Activity들을 보니 1번 표현식으로 적고 있어서 찾아보니 buildGradle에 `databinding = true`를 해주고 layout 태그를 사용했을 때는 1번 표현식으로도 DataBinding이라고 인식한다는 것을 알게 되었습니다.
이전에는 제대로 학습하지 못하고 작성했던 것 같네요... 😢
수정했습니다!

[refactor: DataBindingUtil 사용하지 않도록 변경](https://github.com/woowacourse/android-shopping-order/pull/117/commits/6f66067cca497e4b6a06582c733d00579a7748a2)

krrong commented on app/src/main/java/woowacourse/shopping/feature/cart/CartViewModel.kt:
```diff
@@ -0,0 +1,239 @@
+import woowacourse.shopping.util.toDomain
+import woowacourse.shopping.util.updateQuantity
+
+class CartViewModel(
+    private val cartRepository: CartRepository,
+    private val productRepository: ProductRepository,
+    private val historyRepository: HistoryRepository,
```
조금은 근본적인 질문인데요.
`HistoryRepository`와 `ProductRepository`를 구분한 이유가 있을까요?

우리가 이전 미션에서 데이터 소스를 공부한 이유를 한 번 고민해보면 좋을 것 같습니다!

rosemin928 replied:
`HistoryRepository`와 `ProductRepository`를 구분한 이유는 데이터의 성격과 책임이 다르다고 생각해서였습니다. `ProductRepository`는 서버에서 상품 데이터를 가져오는 역할을 담당하고, `HistoryRepository`는 로컬 저장소에서 사용자의 최근 본 상품이나 행동 기록을 관리하도록 했습니다.

하지만 급하게 구현하며 History에도 Cart를 그대로 넣도록 했었는데, 리팩터링을 진행하며 History 도메인을 따로 생성해 `HistoryRepository`에서는 History에 필요한 정보만 담을 수 있도록 수정해보았습니다!

[refactor: History 도메인 모델 생성 및 관련 로직 수정](https://github.com/woowacourse/android-shopping-order/pull/117/commits/5277ec5f215550b477c79f2137b78cbc6356f332)

krrong replied:
좋은 의견이네요!

제가 생각했던 바는 `ProductRepository`의 이름으로 미루어보면 `Product`를 반환할 것 같아보이죠. 또 `HistoryRepository`는 `History`를 반환할 것 같습니다.

그런데 사실은 `History` 라는 것은 최근 본 상품 즉 `Product`를 반환하는 것으로도 볼 수 있어요. 그렇다면 자연스럽게 `ProductRepository`에서도 그 일을 할 수 있지 않을까? 라는 의견을 드리고 싶었습니다.

`ProductRepository`가 `ProductDatasource`와 `HistoryDatasource`를 각각 가진다면, 뷰모델에서는 최근 본 상품이든, 그냥 상품이든 단순히 `ProductRepository`를 통해서만 가져올 수 있게 되겠죠!

krrong commented on app/src/main/java/woowacourse/shopping/feature/cart/CartViewModel.kt:
```diff
@@ -0,0 +1,239 @@
+            onSuccess = { count ->
+                totalItemsCount.postValue(count.toInt())
+            },
+            onError = { Log.e("loadProductsInRange", "API 요청 실패", it) },
```
보다보니 잊어버리고 지우지 않은 것이 아닌 것 같다는 생각이 드는군요!

rosemin928 replied:
에러 처리에 익숙하지 않아 우선 어떤 에러인지 알 수 있도록 로그를 찍도록 해두었는데 사용자 경험 측면에서는 좋지 않을 것 같다는 생각이 드네요... 🥲
에러가 발생하면 토스트 알림을 띄우는 방식으로 피드백을 주는 방식 정도가 생각이 나는데, 또 이 정도의 방식으로 충분한가?하는 생각도 드네요... 에러 처리에 대한 크롱의 의견이 궁금합니다...! 보통 에러 처리를 어떻게 하시는지 여쭤봐도 될까요??

krrong replied:
예외처리라기 보다는 API요청에 실패한 경우 어떻게 처리하는가로 이야기를 나눠봐도 좋을 것 같아요.

다양한 전략이 있을텐데요.
가장 먼저 비비가 이야기 한 토스트 알림을 띄워줄 수도 있고, 실패한 경우 API를 재요청하도록 하는 방법도 있을 것 같아요. 또는 장바구니 특성상 캐싱을 해서 API 요청에 실패한 경우 캐싱된 정보를 사용할 수도 있겠네요.

로그를 찍는 것은 사실 일반적인 유저 측면에서는 아무런 반응도 없기 때문에 무엇이 문제인지 모르겠죠? 
하지만 일반적이지 않은 유저라면 로그도 확인할 수 있으니 프로덕션 앱에서 로그는 찍지 않습니다 😊

krrong commented on app/src/main/java/woowacourse/shopping/feature/cart/CartViewModel.kt:
```diff
@@ -0,0 +1,239 @@
+        }, 1000) // 스켈레톤 UI 테스트를 위한 딜레이입니다.
+    }
+
+    fun loadProductsByCategory() {
```
콜백의 지옥에 빠지게 되셨네요 😅

rosemin928 replied:
해당 부분은 3단계 코루틴을 적용하며 개선할 수 있을 것 같은데 공부하여 다음 리뷰 반영에 적용해보도록 하겠습니다...!!

krrong commented on app/src/main/java/woowacourse/shopping/feature/cart/CartViewModel.kt:
```diff
@@ -0,0 +1,239 @@
+                                        it.category == category &&
+                                            it.id.toInt() != latestProduct.product.id &&
+                                            it.id !in cartProductIds
+                                    }.take(10)
```
10개를 가져오는 것은 뷰모델이 할 일일까요 리포지토리가 할 일일까요?

rosemin928 replied:
10개를 가져오는 것이 View만을 위한 로직이 아닌, 추천 상품의 비즈니스 로직인 것 같아 레포지토리에서 하는 것이 적절할 것 같습니다!
수정했습니다!

[refactor: 추천 상품을 10개 가져오는 로직을 Repository로 이동](https://github.com/woowacourse/android-shopping-order/pull/117/commits/46969fce5a0d053950d12d546b6c40353c14f9b1)

krrong commented on app/src/main/java/woowacourse/shopping/data/remote/cart/CartClient.kt:
```diff
@@ -0,0 +1,24 @@
+        val contentType = "application/json".toMediaType()
+
+        return Retrofit.Builder()
+            .baseUrl("http://techcourse-lv2-alb-974870821.ap-northeast-2.elb.amazonaws.com")
```
PR 본문에 남겨주신 이 부분은 다음 리뷰 요청까지 한 번 찾아보시면 좋겠네요!

rosemin928 replied:
BuildConfig 자체가 import 되지 않는 문제가 있어서 해결하지 못하고 리뷰 요청을 보내게 되었었는데, 지금 다시 작성해보니 갑자기 되네요...😅
수정했습니다!

[refactor: BuildConfig에서 BASE_URL, USER_ID, USER_PASSWORD 가져오도록 변경](https://github.com/woowacourse/android-shopping-order/pull/117/commits/fc8512a00925cce4806aa8f81f13dec826c46b8c)

krrong replied:
그럴 때가 종종 있죠 😆

krrong commented on app/src/main/java/woowacourse/shopping/data/remote/cart/HeaderInterceptor.kt:
```diff
@@ -0,0 +1,21 @@
+import okhttp3.Interceptor
+import okhttp3.Response
+
+class HeaderInterceptor(
+    private val id: String,
+    private val password: String
+) : Interceptor {
```
필요한 인증 정보를 Interceptor로 잘 활용하고 계시네요 👍

krrong reviewed (changes requested):
안녕하세요 비비! 저도 다시 만나게 되어 영광이네요 😆
새롭고 어려운 내용들이 많았을텐데 너무 잘 구현해주셨습니다!

이번 미션에서는 실제 API와도 연결해보고, 이에 따라 로컬과는 어떻게 연결할 것인지 고민해볼 수 있는데요.
클라이언트 개발자에게는 중요한 내용들이니 이번 기회에 잘 학습해보시면 좋겠어요.

궁금하거나 어려운 점은 언제든 연락주세요! 🙇‍♂️

rosemin928 committed:
refactor: BuildConfig에서 BASE_URL, USER_ID, USER_PASSWORD 가져오도록 변경

rosemin928 committed:
feat: shimmer UI 개선 및 색상 추가

rosemin928 committed:
refactor: MainActivity 및 레이아웃 파일 삭제

rosemin928 committed:
refactor: 로컬 장바구니 관련 파일 제거

- CartDao, CartEntity, CartService, DummyCarts, LocalCartRepository, LocalCartRepositoryImpl 삭제

rosemin928 committed:
refactor: Product 및 HistoryEntity id 타입 Long으로 변경

rosemin928 committed:
feat: 상품 화면의 아이템 목록 sealed class로 변경

- GoodsItem sealed class 정의

rosemin928 committed:
refactor: 상품 목록 RecyclerView Adapter에 ListAdapter 및 DiffUtil 적용

rosemin928 committed:
refactor: History 도메인 모델 생성 및 관련 로직 수정

rosemin928 commented:
@coderabbitai pause

rosemin928 committed:
refactor: 불필요한 domain, feature 모델 삭제

rosemin928 committed:
fix: 최근 본 상품이 하나만 저장되는 오류 수정

rosemin928 committed:
refactor: ViewHolder 생성 로직 companion object로 이동

rosemin928 committed:
fix: 상품 상세페이지에서 수량 변경 시 상품 목록에서 추가/제거가 안 되는 오류 수정

rosemin928 committed:
refactor: Intent extra 키 값 상수화

rosemin928 committed:
refactor: 상품 추가 결과 상태 State 클래스로 관리

rosemin928 committed:
refactor: 장바구니 상품 추가 시 반환값을 Response에서 Cart Id(Long)로 변경

rosemin928 committed:
refactor: ListAdapter와 DiffUtil을 사용하여 RecyclerView 어댑터 개선

rosemin928 committed:
refactor: History 객체 생성 시 명명된 인자 사용

rosemin928 committed:
fix: 장바구니에서 상품 삭제 시 수량 감소 대신 목록에서 제거하도록 수정

rosemin928 committed:
feat: 장바구니 상품 개별 및 전체 선택/해제 기능 추가

rosemin928 committed:
refactor: Cart 도메인 모델명을 CartProduct로 변경

rosemin928 committed:
refactor: GoodsDetailsActivity 로직 정리

rosemin928 committed:
refactor: CartProduct.EMPTY_CART_PRODUCT 상수 추가

rosemin928 committed:
refactor: RecyclerView 아이템 변경 시 깜빡임 현상 수정

rosemin928 committed:
refactor: 뷰홀더에서 클릭 리스너를 init 블록에서 한 번만 설정하도록 변경

rosemin928 committed:
refactor: DataBindingUtil 사용하지 않도록 변경

rosemin928 committed:
refactor: 추천 상품을 10개 가져오는 로직을 Repository로 이동

rosemin928 committed:
refactor: 상품 상세 조회 시 장바구니에 해당 상품이 있는지 확인하는 로직 추가

- 상품 상세 조회 시, 해당 상품이 장바구니에 이미 담겨있는 경우 해당 정보를 함께 불러오도록 수정

rosemin928 committed:
refactor: products와 histories 필드 타입을 MutableList로 변경

rosemin928 committed:
refactor: 최근 본 상품 표시 여부 로직 수정

rosemin928 commented:
안녕하세요 크롱!
생각보다 리뷰 반영이 많이 늦어지게 되었네요... 죄송합니다 🥲
아직도 미흡한 부분이 많지만 이 이상 늦어지게 되면 안 될 것 같아 최대한 반영하고 리뷰 요청 보내드립니다!

언제나 귀한 시간 내어 리뷰해주셔서 감사합니다!! 🙇‍♀️

krrong commented on app/src/main/java/woowacourse/shopping/feature/goods/GoodsActivity.kt:
앱을 처음 로딩하면 가장 상단이 아닌 하단이 보이네요..!

+ 데이터를 더 로딩할 때 전체 상품이 스켈레톤 UI로 변경되었다가 돌아오는 것도 의도된 동작인지 궁금하네요!

https://github.com/user-attachments/assets/f74b3890-fe9e-4c47-8221-56665e264bcf

krrong commented on README.md:
```diff
@@ -1 +1,29 @@
+- [x] 별도의 화면에서 상품 추천 알고리즘으로 사용자에게 적절한 상품을 추천해준다.
+  - [x] 최근 본 상품 카테고리를 기반으로 최대 10개의 상품을 가져온다.
+  - [x] 장바구니에 이미 추가된 상품은 노출하지 않는다.
+- [ ] 추천된 상품을 해당 화면에서 바로 추가하여 같이 주문할 수 있다.
```
요 부분은 무슨 문제가 있는 것일까요?
첨부해주신 영상에서는 잘 실행되는데 제가 실제로 돌려보면 빈 화면만 나오네요 🤔

krrong commented on app/src/main/java/woowacourse/shopping/data/local/ShoppingDatabase.kt:
```diff
@@ -0,0 +1,29 @@
+                    .databaseBuilder(
+                        context.applicationContext,
+                        ShoppingDatabase::class.java,
+                        "app_database",
```
데이터베이스의 이름을 하드코딩하면 어떤 문제가 발생할 수 있을까요?

rosemin928 replied:
데이터베이스의 이름을 하드코딩하면 환경에 따라 다른 DB를 써야 할 때 코드 수정을 반복해야 하고, 이름 변경 시 마이그레이션 누락으로 앱이 크래시될 수도 있다고 하네요…! 우선 상수로 분리하는 방식이 떠올라 분리해보았습니다!

[refactor: 데이터베이스 이름 상수로 추출](https://github.com/woowacourse/android-shopping-order/pull/124/commits/40558b853ec477a98b833b46f90fb700557f537d)

krrong commented on app/src/main/java/woowacourse/shopping/data/remote/product/ProductClient.kt:
```diff
@@ -0,0 +1,23 @@
+import retrofit2.Retrofit
+import woowacourse.shopping.BuildConfig
+
+object ProductClient {
+    fun getRetrofitService(): ProductService {
+        val contentType = "application/json".toMediaType()
+
+        val retrofitService =
+            Retrofit
+                .Builder()
+                .baseUrl(BuildConfig.BASE_URL)
+                .addConverterFactory(Json.asConverterFactory(contentType))
+                .build()
+                .create(ProductService::class.java)
```
`ProductClient`와 `CartClient`는 거의 구현이 동일한데요!
그렇다면 하나의 retrofit 객체를 가지고 재사용해볼 수도 있을 것 같아요!

반영은 하지 않으셔도 되고 참고만 해주셔도 됩니다!

rosemin928 replied:
하나의 retrofit 객체를 재사용하도록 수정했습니다!

[refactor: Product, Cart Client 통합](https://github.com/woowacourse/android-shopping-order/pull/124/commits/b6542a0a9444f4879987ebfb3321b5ba77dbd2c0)

krrong commented on app/src/main/java/woowacourse/shopping/data/remote/cart/CartResponse.kt:
```diff
@@ -0,0 +1,80 @@
+import kotlinx.serialization.SerialName
+import kotlinx.serialization.Serializable
+
+@Serializable
+data class CartResponse(
+    @SerialName("content")
+    val content: List<Content>,
+    @SerialName("empty")
+    val empty: Boolean,
+    @SerialName("first")
+    val first: Boolean,
+    @SerialName("last")
+    val last: Boolean,
+    @SerialName("number")
+    val number: Int,
+    @SerialName("numberOfElements")
+    val numberOfElements: Int,
+    @SerialName("pageable")
+    val pageable: Pageable,
+    @SerialName("size")
+    val size: Int,
+    @SerialName("sort")
+    val sort: Sort,
+    @SerialName("totalElements")
+    val totalElements: Long,
+    @SerialName("totalPages")
+    val totalPages: Int,
+) {
```
이 모든 정보가 필요한가용?.? 다른 객체들도 같은 질문입니다!

rosemin928 replied:
일단 API 응답 구조에 맞춰서 작성했는데, 필요한 정보만 선택적으로 가져와 사용하는 게 더 효율적일 것 같기도 하네요...!
아직 API 로직 경험이 많지 않아 확신이 없네요... 이 부분은 좀 더 학습 및 고민해본 후에 다시 답변을 달아보도록 하겠습니다!! 🙇‍♀️

krrong commented on app/src/main/java/woowacourse/shopping/data/remote/cart/CartResponse.kt:
```diff
@@ -0,0 +1,80 @@
+    val totalPages: Int,
+) {
+    @Serializable
+    data class Content(
```
Response 안에 Content를, Content안에 CartRemoteProduct를 위치시킨 특별한 이유가 있는걸까요?

rosemin928 replied:
응답 데이터의 구조를 그대로 반영하기 위해 CartResponse 안에 Content, Content 내부에 CartRemoteProduct를 넣어 주었습니다! 응답에서 오는 구조를 그대로 사용하면 직렬화 과정에서 오류를 줄일 수 있다고 들어 중첩 구조를 그대로 사용했습니다…!

또, Content가 Response마다 존재하기 때문에 중첩 구조로 넣어줬을 때 

```kotlin
val cartContent: List<CartResponse.Content>
val historyContent: List<HistoryResponse.Content>
```

위와 같이 사용할 수 있어, 좀 더 명확하게 Content를 사용할 수 있겠다고 생각하여 위와 같이 작성했습니다!

krrong commented on app/src/main/java/woowacourse/shopping/data/remote/cart/CartService.kt:
```diff
@@ -0,0 +1,44 @@
+interface CartService {
+    @GET("/cart-items")
+    fun requestCart(
+        @Header("accept") accept: String = "*/*",
```
요 친구는 필요한 이유가 있는건가요~?

rosemin928 replied:
아무 포맷의 응답을 다 받을 수 있도록 적어준 헤더라고 알고 있는데 현재 사용하는 API에서는 필요 없을 것 같네요! 수정했습니다!

[refactor: Retrofit @Header 제거](https://github.com/woowacourse/android-shopping-order/pull/124/commits/a3fd75ee8c945c3f191ca1e3411088fe5b67f3c4)

krrong commented on app/src/main/java/woowacourse/shopping/data/remote/cart/CartRepository.kt:
```diff
@@ -0,0 +1,216 @@
+        onSuccess: (CartResponse) -> Unit,
+        onError: (Throwable) -> Unit,
+    ) {
+        CartClient.getRetrofitService().requestCart(size = Int.MAX_VALUE).enqueue(
```
`CartRepository`에서 API를 호출할 때마다 `Retrofit` 객체를 재생성해주고 있는데요. 어떻게 개선해볼 수 있을지 고민해볼까요?

rosemin928 replied:
하나의 `Retofit` 객체를 재사용할 수 있도록 수정했습니다!
또, `getCartService()`, `getProductService()` 같은 함수들은 이 한 번 생성된 Retrofit 객체를 기반으로 각 API 서비스 인터페이스를 생성해 반환하도록 수정했습니다 😊

[refactor: Product, Cart Client 통합](https://github.com/woowacourse/android-shopping-order/pull/124/commits/b6542a0a9444f4879987ebfb3321b5ba77dbd2c0)

krrong commented on app/src/test/java/woowacourse/shopping/CartViewModelTest.kt:
```diff
@@ -0,0 +1,141 @@
+
+@ExtendWith(InstantTaskExecutorExtension::class)
+@Suppress("ktlint:standard:function-naming")
+class CartViewModelTest {
```
테스트가 많이 깨지고 있네요 😭

krrong commented on app/src/main/java/woowacourse/shopping/feature/cart/CartActivity.kt:
재미있는 오류가 있네요 😆


https://github.com/user-attachments/assets/9fa53a5a-24ab-4141-af9d-a5f9df0ce6eb

rosemin928 replied:
오...int 최대값을 넘어서면 이런 오류가 발생하는군요...! 🤣
totalPrice의 값을 Long으로 변환해주었습니다!

[fix: CartProduct의 totalPrice 자료형을 Int에서 Long으로 변경](https://github.com/woowacourse/android-shopping-order/pull/124/commits/dd901d15723c821b1ba3bc3c1c58dc4a8929a631)

krrong reviewed (approved):
안녕하세요 비비 피드백 반영 너무 잘 해주셨습니다 👍

간단한 코멘트를 몇개 남겨두었으니 다음 단계에서 반영해주시면 됩니다!

마지막 미션의 마지막 단계인만큼 조금만 더 열심히 달려보도록 해요!
궁금하거나 어려운 부분은 언제든 연락주시면 됩니다! 🙇‍♂️
